chia/types/eligible_coin_spends.py

```diff
@@ -97,12 +211,138 @@ def get_deduplication_info(
+            constants: needed in order to refresh the mempool item if needed
+            height: needed in order to refresh the mempool item if needed
+
+        Raises:
+            If a fast forward cannot proceed, to prevent potential double spends
+        """
+        new_coin_spends = []
+        ff_bundle_coin_spends = {}
+        for coin_id, spend_data in mempool_item.bundle_coin_spends.items():
+            if not spend_data.eligible_for_fast_forward:
+                # Nothing to do for this spend, moving on
+                new_coin_spends.append(spend_data.coin_spend)
+                continue
+            # See if we processed a fast forward item with this puzzle hash before
+            unspent_lineage_ids = self.fast_forward_spends.get(spend_data.coin_spend.coin.puzzle_hash)
+            if unspent_lineage_ids is None:
+                # We didn't, so let's lookup the most recent version from the DB
+                unspent_lineage_ids = await get_unspent_lineage_ids_for_puzzle_hash(
+                    spend_data.coin_spend.coin.puzzle_hash
+                )
+                if unspent_lineage_ids is None:
+                    raise ValueError("Cannot proceed with singleton spend fast forward.")
+                # See if we're the most recent version
+                if unspent_lineage_ids.coin_id == coin_id:
+                    # We are, so we don't need to fast forward, we just need to
+                    # set the next version from our additions to chain ff spends
+                    set_next_singleton_version(
+                        current_singleton=spend_data.coin_spend.coin,
+                        singleton_additions=spend_data.additions,
+                        fast_forward_spends=self.fast_forward_spends,
+                    )
+                    # Nothing more to do for this spend, moving on
+                    new_coin_spends.append(spend_data.coin_spend)
+                    continue
+                # We're not the most recent version, so let's fast forward
+                new_coin_spend, patched_additions = perform_the_fast_forward(
+                    unspent_lineage_ids=unspent_lineage_ids,
+                    spend_data=spend_data,
+                    fast_forward_spends=self.fast_forward_spends,
+                )
+                # Mark this coin for a coin spend data update
+                ff_bundle_coin_spends[coin_id] = BundleCoinSpend(
+                    coin_spend=new_coin_spend,
+                    eligible_for_dedup=spend_data.eligible_for_dedup,
+                    eligible_for_fast_forward=spend_data.eligible_for_fast_forward,
+                    additions=patched_additions,
+                    cost=spend_data.cost,
+                )
+                # Update the list of coins spends that will make the new fast
+                # forward spend bundle
+                new_coin_spends.append(new_coin_spend)
+                # We're done here, moving on
```

Comment:
this looks very similar to the section below, outside of this if-block. does the code end up more complex if you try to join this code with the one below?

Reply:
Yep, I originally explored different styles, but this is the most beneficial in terms of readability and maintainability.